Add API for consuming solution and document-specific options

jasonmalinowski opened the pull request:
This is a large change to introduce an API to consume document- and project-level options. Just like `Workspace.Options` today gives you an `OptionSet`, this introduces `Document.Options` and `Solution.Options`. Right now they just return the workspace-level values, but this gives us the ability to make those do something else in upcoming days/weeks.

*Review:* @dotnet/roslyn-ide, @dotnet/roslyn-analysis

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/LanguageService/AbstractLanguageService`3.cs:
```diff
@@ -144,7 +143,7 @@ protected override void SetupNewTextView(IVsTextView textView)
         }
 
-        private void ConditionallyCollapseOutliningRegions(IVsTextView textView, IWpfTextView wpfTextView, IOptionService optionsService, bool isOpenMetadataAsSource)
+        private void ConditionallyCollapseOutliningRegions(IVsTextView textView, IWpfTextView wpfTextView, VisualStudioWorkspace workspace, bool isOpenMetadataAsSource)
```
Can you just take Workspace isntead?

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Workspace/Solution/Document.cs:
```diff
@@ -425,5 +426,18 @@ private string GetDebuggerDisplay()
+        /// Returns the options that should be applied to this document. This consists of global options from <see cref="Workspace.Options"/>,
+        /// merged with any settings the user has specified at the solution, project, and document levels.
+        /// </summary>
+        public OptionSet Options
```
Use => form. :)

CyrusNajmabadi replied:
Ah... except you intend to eventually do merging.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs:
```diff
@@ -2283,5 +2284,18 @@ private void CheckContainsAdditionalDocument(DocumentId documentId)
+            get
+            {
+                // TODO: merge with solution-specific options
+                return this.Workspace.Options;
```
Don't like that this means Solution.Options is observably mutable.

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/ExtractMethod/AbstractExtractMethodCommandHandler.cs:
```diff
@@ -224,8 +223,10 @@ private bool TryNotifyFailureToUser(Document document, ExtractMethodResult resul
 
         private static ExtractMethodResult TryWithoutMakingValueTypesRef(
-            Document document, NormalizedSnapshotSpanCollection spans, OptionSet options, ExtractMethodResult result, CancellationToken cancellationToken)
+            Document document, NormalizedSnapshotSpanCollection spans,  ExtractMethodResult result, CancellationToken cancellationToken)
```
Extra space after comma

CyrusNajmabadi commented on src/Features/Core/Portable/CodeRefactorings/ExtractMethod/AbstractExtractMethodCodeRefactoringProvider.cs:
```diff
@@ -59,17 +59,16 @@ protected AbstractExtractMethodCodeRefactoringProvider()
                 textSpan,
-                options,
-                cancellationToken).ConfigureAwait(false);
+                options: null,
```
is there an overlaod that doesn't take options?  Feels like that one should eb called.  Also, if this is an internal method, we should just remove that parameter.  If it is public we should obsolete this version (IMO).

CyrusNajmabadi commented:
:+1: 

As mentioned in our talks:

1. I would like DOcument/Project/Solution/.Options to be observably immutable.  I think we get this by making it so that only Workspace.Options is mutable.  When workspace options change you get a new solution snapshot.
2. The options you get from Document/Project should be specialized so that you don't have to mention the language when getting PerLanguageOptions.  Maybe this is achieved through returning some subclass of OptionSet that has helper overloads of these methods.

Overall though, this looks nice and i'm glad this saved us 150 LOC.

heejaechang commented:
+1 on what @CyrusNajmabadi said.

jasonmalinowski commented:
I've updated this to fix @CyrusNajmabadi's issue #2. The immutable problem "already exists", and I'll fix that once we actually have the system that provides document settings, since that has to have some caching logic in it that will fix it. You can re-review commit 2550bb0c01f49e290c6f9f74cb97aabf6a05cbc5 to see just what that looks like.

I also removed Project.Options, since I couldn't really find a reasonable use for it in the end. You *don't* want to use it for formatting options since formatting options can be per-document. Since most uses would be wrong and none would be right, I removed it for now.

@CyrusNajmabadi or @heejaechang, can you re-review 2550bb0c01f49e290c6f9f74cb97aabf6a05cbc5? Also @jaredpar or @davkean I would like another look to actually confirm that I didn't screw up the non-breaking-changeness of this.

jasonmalinowski commented:
@dotnet-bot, test vsi please.

jasonmalinowski commented:
@dotnet-bot, test vsi please.

jasonmalinowski commented:
@dotnet-bot, retest roslyn_prtest_win_vsi0 please. This appears to be a duplicate of #10916.

jasonmalinowski commented:
@dotnet-bot, retest roslyn_prtest_win_vsi0 please. This appears to be a duplicate of #10916.

jasonmalinowski commented:
@dotnet-bot, test vsi please.

jasonmalinowski commented:
@dotnet-bot, test vsi please.

jasonmalinowski commented:
@dotnet-bot, test vsi please.

jaredpar commented:
test vsi please

jaredpar commented:
test vsi please

jaredpar commented:
test vsi please

jasonmalinowski commented:
test vsi please

davkean commented:
test vsi please

jaredpar commented:
test vsi please
